base_layer/wallet/tests/transaction_service/transaction_protocols.rs

```diff
@@ -283,9 +283,10 @@ async fn tx_broadcast_protocol_submit_success() {
 
     // Wait for 2 queries
     let _ = rpc_service_state
-        .wait_pop_submit_transaction_calls(4, Duration::from_secs(5))
+        .wait_pop_submit_transaction_calls(4, Duration::from_secs(6))
         .await
         .unwrap();
+    tokio::time::delay_for(Duration::from_secs(1)).await;
```

Comment:
I don't think these delays after the  `wait_pop_submit_transaction_calls` do anything useful, the `wait_pop_submit_transaction_calls` are already waiting?

Reply:
So what happens is that call to read the database gets in before the call to write the database. 
The Tx manager already has the info, but it has not yet saved it to the db. That's why this small sleep solves this. When it fails you can see its because the read tx got its read lock before the write tx could get it.

Reply:
I don't see how this solves the problem you are stating, the two statements you are putting the delay between only change the response returned to the transaction protocol when it makes a query, it has nothing to do with a database? The transaction protocol is running in its own task so this delay has no effect on it. The thing this is delaying between is changing the response the RPC mock gives to the transaction protocol when it makes its RPC request. By adding this delay you actually make it that the transaction protocol will probably put in a second request before the response type changes (the protocol is set to poll every 1 second).

Reply:
I think this time_delay does not help and might be redundant. The next ones should give time for the DB to write its changes. 

But after this discussion, it's better to move to an event-based fix. I have added an event listen here to wait for the db write.